src/GafferScene/Constraint.cpp

```diff
@@ -38,11 +38,594 @@
+	if( pdata == 0 )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( "Gaffer::Contraint : Primitive has no Vertex \"P\" primitive variable." );
+		}
+		else
+		{
+			return;
+		}
+	}
+	const IECore::V3fVectorData::ValueType& points = pdata->readable();
+
+	if( ( vertexId < 0 ) || ( vertexId >= static_cast< int >( points.size() ) ) )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( (
+				boost::format(
+					"Gaffer::Constraint : Vertex id \"%d\" is out of range."
+				) % ( vertexId ) ).str() );
+		}
+		else
+		{
+			return;
+		}
+	}
+
+	m.translate( points[ vertexId ] );
+}
+
+void computeMeshVertexLocalFrame( const IECoreScene::MeshPrimitive& primitive, Imath::M44f& m, const int vertexId, const std::string& uvSet, const bool throwOnError, const IECore::Canceller* const canceller )
+{
+	const IECore::V3fVectorData* const pdata = primitive.variableData< IECore::V3fVectorData >( "P" );
+
+	if( pdata == 0 )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( "Gaffer::Contraint : MeshPrimitive has no Vertex \"P\" primitive variable." );
+		}
+		else
+		{
+			return;
+		}
+	}
+
+	const IECore::V3fVectorData::ValueType& points = pdata->readable();
+
+	if( ( vertexId < 0 ) || ( vertexId >= static_cast< int >( points.size() ) ) )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( (
+				boost::format(
+					"Gaffer::Constraint : Vertex id \"%d\" is out of range."
+				) % ( vertexId ) ).str() );
+		}
+		else
+		{
+			return;
+		}
+	}
+
+	const IECore::IntVectorData::ValueType& indices = primitive.vertexIds()->readable();
+	const IECore::IntVectorData::ValueType& faces = primitive.verticesPerFace()->readable();
+
+	const UVIndexer uvs( primitive, uvSet, throwOnError );
+
+	Imath::V3f t( 0.f );
+	Imath::V3f b( 0.f );
+
+	if( uvs.valid() )
+	{
+		for( int i = 0, nf = faces.size(), js = 0; i < nf; ++i )
+		{
+			const int ni = faces[ i ];
+
+			// canceller support
+
+			if( ( i % 100 ) == 0 )
+			{
+				IECore::Canceller::check( canceller );
+			}
+
+			// find matching face vertex index
+
+			int jm = 0;
+			for( ; jm < ni; ++jm )
+			{
+				if( indices[ js + jm ] == vertexId )
+				{
+					break;
+				}
+			}
+
+			if( jm < ni )
+			{
+				// compute face tangent and bitangent
+
+				Imath::V3f ft( 0.f );
+				Imath::V3f fb( 0.f );
+
+				float w = 0.f;
+
+				for( int j = 0; j < ni; ++j )
+				{
+					const int iv = js + j;
+					const int ip = js + ( ( j + ni - 1 ) % ni );
+					const int in = js + ( ( j      + 1 ) % ni );
+
+					const Imath::V3f pv = points[ indices[ iv ] ];
+					const Imath::V3f v0 = points[ indices[ ip ] ] - pv;
+					const Imath::V3f v2 = points[ indices[ in ] ] - pv;
+
+					const Imath::V2f uv = uvs[ iv ];
+					const Imath::V2f e0 = uvs[ ip ] - uv;
+					const Imath::V2f e2 = uvs[ in ] - uv;
+
+					ft += ( v0 * -e2.y + v2 * e0.y ).normalized();
+					fb += ( v0 * -e2.x + v2 * e0.x ).normalized();
+
+					if( j == jm )
```

Comment:
Terseness of variables is an interesting question ... I used to be a lot more terse before I got to IE, and adjusted to abbreviating a lot less, and may have gone too far now.  I don't mind a lot of these two letter variables which are used quite close to being declared, but something like "jm" is being used far enough away from its declaration that maybe something way more explicit like "matchingIndex" could be called for ... I dunno, when it's internal to a function it really doesn't matter too much.